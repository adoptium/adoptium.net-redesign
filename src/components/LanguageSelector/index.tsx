import React, { Fragment, useState } from "react"
import { useI18next } from "gatsby-plugin-react-i18next"
import { Menu, MenuButton, MenuItem, MenuItems, Transition } from "@headlessui/react"
import { FaChevronDown } from "react-icons/fa"
import Flag from "react-world-flags"
import ISO6391 from "iso-639-1"

function classNames(...classes) {
  return classes.filter(Boolean).join(" ")
}

const LanguageSelector = (): JSX.Element => {
  const { languages, changeLanguage, language } = useI18next()
  const [selectedLanguage, setSelectedLanguage] = useState(language)

  function ISO3166(lng: string) {
    // Convert locale to ISO 3166-1 alpha-2 https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2
    switch (lng) {
      case "en":
        return "us"
      case "en-GB":
        return "gb"
      case "zh-CN":
        return "cn"
      default:
        return lng
    }
  }

  function ISO639(lng: string) {
    // Convert locale to ISO 639-1 alpha-2 https://en.wikipedia.org/wiki/ISO_639-1_alpha-2
    switch (lng) {
      case "zh-CN":
        return "zh"
      case "en-GB":
        return "en"
      default:
        return lng
    }
  }

  return (
    <Menu as="div" className="relative inline-block text-left">
      <div>
        <MenuButton className="inline-flex w-full justify-center gap-3 rounded-3xl bg-transparent h-full px-4 py-3 text-sm text-white shadow-xs border-2 border-gray-700">
          <Flag className="mb-0 h-5" code={ISO3166(language)} />
          <span>{ISO6391.getNativeName(ISO639(selectedLanguage))}</span>
          <FaChevronDown
            className="-mr-1 h-5 w-5 text-white"
            aria-hidden="true"
            size={15}
          />
        </MenuButton>
      </div>

      <Transition
        as={Fragment}
        enter="transition ease-out duration-100"
        enterFrom="transform opacity-0 scale-95"
        enterTo="transform opacity-100 scale-100"
        leave="transition ease-in duration-75"
        leaveFrom="transform opacity-100 scale-100"
        leaveTo="transform opacity-0 scale-95"
      >
        <MenuItems className="absolute right-0 z-10 mt-2 w-32 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black/5 focus:outline-hidden">
          <div className="py-1">
            {languages.map((lng: string) => (
              <MenuItem key={lng}>
                {({ focus }) => (
                  <a
                    href="#"
                    id={lng}
                    data-testid={lng}
                    onClick={e => {
                      e.preventDefault()
                      changeLanguage(lng)
                    }}
                    className={classNames(
                      focus ? "bg-gray-100 text-gray-900" : "text-gray-700",
                      "block px-4 py-2 text-sm",
                    )}
                  >
                    <div className="flex items-center space-x-2">
                      <Flag className="mb-0" code={ISO3166(lng)} width="35" />
                      <span className="ml-2">{ISO6391.getNativeName(ISO639(lng))}</span>
                    </div>
                  </a>
                )}
              </MenuItem>
            ))}
          </div>
        </MenuItems>
      </Transition>
    </Menu>
  )
}

export default LanguageSelector
