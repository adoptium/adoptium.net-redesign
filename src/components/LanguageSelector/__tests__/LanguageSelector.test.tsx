import React from "react"
import { render, screen, fireEvent } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { describe, expect, it, vi } from "vitest"
import LanguageSelector from ".."

window.ResizeObserver = vi.fn().mockImplementation(() => ({
  observe: vi.fn(),
  unobserve: vi.fn(),
  disconnect: vi.fn(),
}))

describe("Language Selector component", () => {
  it("renders correctly", () => {
    const { container } = render(<LanguageSelector />)
    expect(container).toMatchSnapshot()
  })

  it("renders correctly - expanded", async () => {
    const { container } = render(<LanguageSelector />)
    const dropdownButton = screen.getByRole("button")
    await userEvent.click(dropdownButton).then(async () => {
      expect(container).toMatchSnapshot()
      // Simulate changing the language using the dropdown
      const dropDownElement = screen.getByTestId("en-GB")
      fireEvent.click(dropDownElement)
    })
  })
})
