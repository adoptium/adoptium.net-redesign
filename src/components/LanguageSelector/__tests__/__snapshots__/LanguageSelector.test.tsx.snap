// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Language Selector component > renders correctly - expanded 1`] = `
<div>
  <div
    class="relative inline-block text-left"
    data-headlessui-state="open"
    data-open=""
  >
    <div>
      <button
        aria-controls="headlessui-menu-items-:r6:"
        aria-expanded="true"
        aria-haspopup="menu"
        class="inline-flex w-full justify-center gap-3 rounded-3xl bg-transparent h-full px-4 py-3 text-sm text-white shadow-xs border-2 border-gray-700"
        data-active=""
        data-headlessui-state="open active hover"
        data-hover=""
        data-open=""
        id="headlessui-menu-button-:r5:"
        type="button"
      >
        Flag
        <span>
          English
        </span>
        <svg
          aria-hidden="true"
          class="-mr-1 h-5 w-5 text-white"
          fill="currentColor"
          height="15"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
          />
        </svg>
      </button>
    </div>
    <div
      aria-labelledby="headlessui-menu-button-:r5:"
      class="absolute right-0 z-10 mt-2 w-32 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black/5 focus:outline-hidden transition ease-out duration-100 transform opacity-0 scale-95"
      data-closed=""
      data-enter=""
      data-headlessui-state="open"
      data-open=""
      data-transition=""
      id="headlessui-menu-items-:r6:"
      role="menu"
      style="--button-width: 0px;"
      tabindex="0"
    >
      <div
        class="py-1"
        role="none"
      >
        <a
          class="text-gray-700 block px-4 py-2 text-sm"
          data-headlessui-state=""
          data-testid="en"
          href="#"
          id="headlessui-menu-item-:r9:"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="flex items-center space-x-2"
          >
            Flag
            <span
              class="ml-2"
            >
              English
            </span>
          </div>
        </a>
        <a
          class="text-gray-700 block px-4 py-2 text-sm"
          data-headlessui-state=""
          data-testid="en-GB"
          href="#"
          id="headlessui-menu-item-:ra:"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="flex items-center space-x-2"
          >
            Flag
            <span
              class="ml-2"
            >
              English
            </span>
          </div>
        </a>
        <a
          class="text-gray-700 block px-4 py-2 text-sm"
          data-headlessui-state=""
          data-testid="es"
          href="#"
          id="headlessui-menu-item-:rb:"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="flex items-center space-x-2"
          >
            Flag
            <span
              class="ml-2"
            >
              Español
            </span>
          </div>
        </a>
        <a
          class="text-gray-700 block px-4 py-2 text-sm"
          data-headlessui-state=""
          data-testid="de"
          href="#"
          id="headlessui-menu-item-:rc:"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="flex items-center space-x-2"
          >
            Flag
            <span
              class="ml-2"
            >
              Deutsch
            </span>
          </div>
        </a>
        <a
          class="text-gray-700 block px-4 py-2 text-sm"
          data-headlessui-state=""
          data-testid="zh-CN"
          href="#"
          id="headlessui-menu-item-:rd:"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="flex items-center space-x-2"
          >
            Flag
            <span
              class="ml-2"
            >
              中文
            </span>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Language Selector component > renders correctly 1`] = `
<div>
  <div
    class="relative inline-block text-left"
    data-headlessui-state=""
  >
    <div>
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        class="inline-flex w-full justify-center gap-3 rounded-3xl bg-transparent h-full px-4 py-3 text-sm text-white shadow-xs border-2 border-gray-700"
        data-headlessui-state=""
        id="headlessui-menu-button-:r2:"
        type="button"
      >
        Flag
        <span>
          English
        </span>
        <svg
          aria-hidden="true"
          class="-mr-1 h-5 w-5 text-white"
          fill="currentColor"
          height="15"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
